In [2]:
import os
import json
import warnings
import numpy as np
import pandas as pd
import seaborn as sns
from scripts import *
from tqdm import tqdm
import plotly.io as pio
import plotly.express as px
import matplotlib.pyplot as plt
warnings.filterwarnings('ignore')
pd.set_option('display.max_columns', 200)
%matplotlib inline
api_key = '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'

In [4]:
data_dir = os.path.abspath("../data")
results_dir = os.path.abspath("../results")

In [13]:
fb_comments = f'{data_dir}/webscrape_fbhashtag_LM_comments-2.csv'
df_comments = pd.read_csv(fb_comments)

In [14]:
df_comments.head()

date                                               text
0  2021-05-30T02:34:03.000Z  Wow sarap tlaga bsta pancit canton may fav chi...
1  2021-05-30T05:39:15.000Z  Favorite namen ng kids ko yan momsh Lalo ng pa...
2  2021-05-30T03:28:16.000Z  Favorite merienda namin to lalo na ang pancit ...
3  2021-05-30T10:27:27.000Z  I remember the days na ito yung nagpapa surviv...
4  2021-05-31T10:32:35.000Z  Sarap fav.namin ng mga anak ko pero my limit p...

In [15]:
df_comments.shape

(2119, 2)

In [16]:
df_comments.dtypes

date    object
text    object
dtype: object

In [17]:
df_comments['date'] =pd.to_datetime(df_comments['date']).dt.date
df_comments.head()

date                                               text
0  2021-05-30  Wow sarap tlaga bsta pancit canton may fav chi...
1  2021-05-30  Favorite namen ng kids ko yan momsh Lalo ng pa...
2  2021-05-30  Favorite merienda namin to lalo na ang pancit ...
3  2021-05-30  I remember the days na ito yung nagpapa surviv...
4  2021-05-31  Sarap fav.namin ng mga anak ko pero my limit p...

In [18]:
os.environ["OPENAI_API_KEY"] = api_key

from langchain.chat_models import ChatOpenAI

chat = ChatOpenAI(temperature=0.4)

- Exception - need to delete rows 1738 to 1749 due to connection issues in processing the sentiments
- I tried re-running 8x but it always gets disconnected. OpenAI credits spent

In [19]:
df_comments.drop(df_comments.index[1739:1750], axis=0, inplace=True)

In [20]:
df_comments.shape

(2108, 2)

In [21]:
df_comments.to_csv(f'{data_dir}/fbcomments_cleaned.csv')

### Prompt Template + LLM

In [9]:
fbcomment_template = """\
For the following Facebook comment, extract the following information:

topic: What is the topic of the comment? Give 1-2 topics in list format. \
If there is a mention of "LuckyMe" or any other brand, include it.

sentiment: What is the sentiment of the comment? \
Answer positive, negative or neutral.

summary: Summarize the post in 1-2 sentences.

product_rating: What does the customer say about the quality of the \
product based on his/her comment? If there is a mention of product quality, 
give it a rating on a scale of 1 to 5, 5 being the highest." \
"Otherwise, if there is no mention of product quality, based on the sentiment, \
on a scale of 1 to 5, 5 being the highest, what product quality rating can you give?"

Format the output as JSON with the following keys:
topic
sentiment
summary
product_rating

post: {comment}
"""

from langchain.prompts import ChatPromptTemplate

prompt_template = ChatPromptTemplate.from_template(fbcomment_template)

In [10]:
prompt_template.input_variables

['comment']

### Divide into chunks of data due to connection issues

In [11]:
fb = df_comments.drop(columns='date', axis=1).copy()
fb

text
0     Wow sarap tlaga bsta pancit canton may fav chi...
1     Favorite namen ng kids ko yan momsh Lalo ng pa...
2     Favorite merienda namin to lalo na ang pancit ...
3     I remember the days na ito yung nagpapa surviv...
4     Sarap fav.namin ng mga anak ko pero my limit p...
...                                                 ...
2114                                              Junot
2115                                                NaN
2116                 Charlie Canda hehehe mao bitaw pre
2117               June Handugan empe ta junot pre hehe
2118  paki pump slp Jeff Jee Ho Zirlin wag ka maglut...

[2119 rows x 1 columns]

In [17]:
# convert to list
fb_list = list(fb['text'])

In [18]:
# Initialize an empty list to store sublists
sublists = []

# Define the size of each sublist
sublist_size = 250

# Use a loop to create sublists
for i in range(0, len(fb_list), sublist_size):
    sublist = fb_list[i:i + sublist_size]
    sublists.append(sublist)

In [24]:
sublists[1]

['These should be mainstream flavors! 🧡',
 'Finally! I’ve been mixing these flavors for years! At least 1 nalang bibilhin ko nyan 😍',
 'Wow sarap 😍',
 'ako lang ba at mga anak ko na ang gusto sa noodless ng pancit canton eh medyo durog pag niluto. 😅😅😅',
 'All time favorite. 👌',
 'Ang perfect ng egg 😍',
 'Shuuucks bawal na ko sa maaalat eh ang sherep 😍',
 'Sweet n spicy chili mansi!!!!??',
 'Sweet spicyMansi 😋😍',
 'Gabi na kaseeee',
 '🙀🙀🙀',
 'Mark Onato',
 '🧡',
 'Parang gusto ko nito. 🙈 \nPagluto nyo ko pagbalik ko. 🤪 \nLheny Paclipan Garcia Jackie Balaba',
 'DEx Donato',
 'Huy Micko Advincula',
 'Sana all Kevz S. Viray',
 "Tignan n'yo yung sodium content nyan sa sodium content ng isang boteng toyo. Diba para ka na ring tumaklak ng isang boteng toyo 2x! Kaya hinay-hinay lang sa pagkain n'yan. Good luck!",
 'Grabe naaaa Kristine Marel Juan Javier',
 'Blaire Raphael Salamat',
 'Eto na lunch naten mamaya Natalia Negrito',
 'Jasmine Tamayo, oras na para bumalik. 🤩',
 'Lucky Me pancit canton

### Prompt Template +LLM + Output Parser

In [21]:
from langchain.output_parsers import ResponseSchema
from langchain.output_parsers import StructuredOutputParser

topic_schema = ResponseSchema(
    name="topic",
    description="What is the main topic of the post. If there are more than 1 topic, \
        do not put in [] instead, separate them with comma. \
        If there is a mention of 'LuckyMe' or any other brand, include it."
)

sentiment_schema = ResponseSchema(
    name="sentiment",
    description="What is the sentiment of the post? \
        Answer positive, negative or neutral."
)

summary_schema = ResponseSchema(
    name="summary",
    description="Summarize the post."
)

product_rating_schema= ResponseSchema(
    name="product_rating",
    description="What does the customer say about the quality of the \
        product based on his/her post? If there is a mention of product quality, \
        give it a rating on a scale of 1 to 5, 5 being the highest. \
        Otherwise, if there is no mention of product quality, based on the sentiment, \
        on a scale of 1 to 5, 5 being the highest, what product quality rating can you give? \
        If there is no relevance to the product, give it a rating of NA"
)

response_schemas = [
    topic_schema,
    sentiment_schema,
    summary_schema,
    product_rating_schema
]

In [ ]:
# def json_dict_df_csv(suffx, responses_list):
#     # Initialize an empty list to store dictionaries
#     dict_list = []

#     # Parse each JSON string and convert it to a dictionary
#     for json_str in responses_list:
#         data_dict = json.loads(json_str)
#         dict_list.append(data_dict)

#     # Create a DataFrame from the list of dictionaries
#     fbcomment_df = pd.DataFrame(dict_list)
#     print('Saving to CSV file...')
    # fbcomment_df.to_csv(f'{data_dir}/sentiment_analysis_fbposts'+str(suffx)+'.csv', index=False)

In [15]:
output_parser = StructuredOutputParser.from_response_schemas(response_schemas)
format_instructions = output_parser.get_format_instructions()

In [ ]:
responses_list=[]

for text_comment in tqdm(fb_list, desc="Processing"):

    #for text in fb_comments:
    messages = prompt_template.format_messages(
        comment=text_comment,
        format_instructions=format_instructions
        )
    response = chat(messages)
    output_dict = output_parser.parse(response.content)
    responses_list.append(response.content)
    print(response.content)

    pass

In [ ]:
# Initialize an empty list to store dictionaries
dict_list = []

# Parse each JSON string and convert it to a dictionary
for json_str in responses_list:
    data_dict = json.loads(json_str)
    dict_list.append(data_dict)

# Create a DataFrame from the list of dictionaries
fbcomment_df = pd.DataFrame(dict_list)

fbcomment_df.to_csv(f'{data_dir}/sentiment_analysis_fbcomments-1.csv', index=False)

In [32]:
x = pd.read_csv(f'{data_dir}/fbcomments_for_viz.csv')
x.shape

(2108, 6)

In [33]:
x.head()

date                                            comment  \
0  2021-05-30  Wow sarap tlaga bsta pancit canton may fav chi...   
1  2021-05-30  Favorite namen ng kids ko yan momsh Lalo ng pa...   
2  2021-05-30  Favorite merienda namin to lalo na ang pancit ...   
3  2021-05-30  I remember the days na ito yung nagpapa surviv...   
4  2021-05-31  Sarap fav.namin ng mga anak ko pero my limit p...   

                                    topic sentiment  \
0               ['Food', 'Pancit Canton']  positive   
1  ['Favorite names for kids', 'LuckyMe']  positive   
2           ['Merienda', 'Pancit Canton']  positive   
3                ['College days', 'Work']  positive   
4                     ['Food', 'LuckyMe']  positive   

                                             summary  product_rating  
0  The commenter expresses delight in the taste o...               5  
1  The commenter's favorite names for their kids ...               5  
2  The post is about the user's favorite merienda...               5  
3  The commenter reminisces about how this produc...               5  
4  The commenter enjoys the product and their chi...               5

In [34]:
x['sentiment'].value_counts()

neutral     1034
positive     974
negative      68
Neutral       26
Unknown        3
Positive       2
unknown        1
Name: sentiment, dtype: int64

### Files:
- fbcomments_cleaned.csv = 2108 original comments with date
- fbcomments_for_viz.csv = 2108 processed, with sentiments and topic
- fbpost_cleaned.csv = 515 original post with other features
- fbpost_for_viz.csv = 515 processed, with sentiments and topic
